src/block/CaveVines.php

```diff
@@ -0,0 +1,203 @@
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\block;
+
+use pocketmine\block\utils\CaveVinesType;
+use pocketmine\block\utils\SupportType;
+use pocketmine\data\runtime\RuntimeDataReader;
+use pocketmine\data\runtime\RuntimeDataWriter;
+use pocketmine\entity\Entity;
+use pocketmine\event\block\BlockGrowEvent;
+use pocketmine\item\Fertilizer;
+use pocketmine\item\Item;
+use pocketmine\item\VanillaItems;
+use pocketmine\math\Facing;
+use pocketmine\math\Vector3;
+use pocketmine\player\Player;
+use pocketmine\world\BlockTransaction;
+use pocketmine\world\sound\GlowBerriesPickSound;
+use function mt_rand;
+
+class CaveVines extends Flowable{
+	public const MAX_AGE = 25;
+
+	protected CaveVinesType $type;
+
+	protected int $age = 0;
+
+	public function __construct(BlockIdentifier $idInfo, string $name, BlockTypeInfo $typeInfo){
+		$this->type = CaveVinesType::BODY();
+		parent::__construct($idInfo, $name, $typeInfo);
+	}
+
+	public function getRequiredTypeDataBits() : int{
+		return 2;
+	}
+
+	public function describeType(RuntimeDataWriter|RuntimeDataReader $w) : void{
+		$w->caveVinesType($this->type);
+	}
+
+	public function getRequiredStateDataBits() : int{
+		return 5;
+	}
+
+	public function describeState(RuntimeDataWriter|RuntimeDataReader $w) : void{
+		$w->boundedInt(5, 0, self::MAX_AGE, $this->age);
+	}
+
+	public function getType() : CaveVinesType{
+		return $this->type;
+	}
+
+	/** @return $this */
+	public function setType(CaveVinesType $type) : self{
+		$this->type = $type;
+		return $this;
+	}
+
+	public function getAge() : int{
+		return $this->age;
+	}
+
+	/** @return $this */
+	public function setAge(int $age) : self{
+		if($age < 0 || $age > self::MAX_AGE){
+			throw new \InvalidArgumentException("Age must be in range 0-" . self::MAX_AGE);
+		}
+		$this->age = $age;
+		return $this;
+	}
+
+	public function isAffectedBySilkTouch() : bool{
+		return true;
+	}
+
+	public function ticksRandomly() : bool{
+		return true;
+	}
+
+	public function canClimb() : bool{
+		return true;
+	}
+
+	public function getLightLevel() : int{
+		return $this->type->hasBerries() ? 14 : 0;
+	}
+
+	private function canBeSupportedBy(Block $block) : bool{
+		return $block->getSupportType(Facing::DOWN)->hasCenterSupport() || $block->isSameType($this);
```

Comment:
I would expect that HEAD_WITH_BERRIES cannot provide support to other vine blocks. That should probably be checked here.

Reply:
I re-tested this behavior in vanilla and it's more confusing than i intend. In vanilla, HEAD_WITH_BERRIES can provide support to other vines, and vanilla recalculates the type when a block has been ticked randomly, not when some block has been changed nearby